classes/local/step/compression_trait.php

```diff
@@ -0,0 +1,267 @@
+
+        $mform->addElement('select', 'config_method', get_string('compression:method', 'tool_dataflows'), $methodoptions);
+
+        // From / Source path.
+        $mform->addElement('text', 'config_from', get_string('compression:from', 'tool_dataflows'));
+        $mform->addRule('config_from', get_string('required'), 'required', null, 'client');
+
+        // To / Target path.
+        $mform->addElement('text', 'config_to', get_string('compression:to', 'tool_dataflows'));
+        $mform->addRule('config_to', get_string('required'), 'required', null, 'client');
+    }
+
+    /**
+     * Executes the step
+     *
+     * @param  mixed|null $input
+     * @return mixed
+     */
+    public function execute($input = null) {
+        $variables = $this->get_variables();
+        $config = $variables->get('config');
+        $config->from = $this->enginestep->engine->resolve_path($config->from);
+        $config->to = $this->enginestep->engine->resolve_path($config->to);
+
+        // Check that the from path exists.
+        if (!is_file($config->from)) {
+            $this->enginestep->log($config->from . ' file does not exist');
+            $variables->set('vars.success', false);
+            return $input;
+        }
+
+        // We do not need to go any further if it is a dry run.
+        if ($this->is_dry_run() && $this->has_side_effect()) {
+            return $input;
+        }
+
+        $result = $this->execute_method($config);
+
+        if ($result !== true) {
+            // Log the error.
+            $this->enginestep->log($result);
+        }
+
+        $variables->set('vars.success', $result === true);
+
+        return $input;
+    }
+
+    /**
+     * Perform any extra validation that is required only for runs.
+     *
+     * @return true|array Will return true or an array of errors.
+     */
+    public function validate_for_run() {
+        $config = $this->get_variables()->get('config');
+
+        $errors = [];
+
+        $error = helper::path_validate($config->from);
+        if ($error !== true) {
+            $errors['config_from'] = $error;
+        }
+
+        $error = helper::path_validate($config->to);
+        if ($error !== true) {
+            $errors['config_to'] = $error;
+        }
+
+        // Valid the chosen methods executable is actually executable.
+        $method = $this->get_method($config->method);
+        $error = ($method->isexecutable)();
+        if ($error !== true) {
+            $errors['config_method'] = $error;
+        }
+
+        return $errors ?: true;
+    }
+
+    /**
+     * A list of outputs and their description if applicable.
+     *
+     * @return  array of outputs
+     */
+    public function define_outputs(): array {
+        return ['success' => get_string('compression:output_success', 'tool_dataflows')];
+    }
+
+    /**
+     * Returns the method info that has been selected in the configuration.
+     * @param string $method method name
```

Comment:
It would appear more consistent to space the doc summary and parameters.

```suggestion
     * Returns the method info that has been selected in the configuration.
     *
     * @param string $method method name
```

This would apply to other similar places in this PR too.

Reply:
Good catch - have updated these to be consistent :+1: